runtime/oti/VMHelpers.hpp

```diff
@@ -648,13 +648,18 @@ class VM_VMHelpers
 							/* check the [[O -> [[O case.  Don't allow [[I -> [[O */
 							if (instanceArity == castArity) {
 								J9Class *instanceClassLeafComponent = ((J9ArrayClass*)instanceClass)->leafComponentType;
-								if (J9CLASS_IS_MIXED(instanceClassLeafComponent)) {
-									/* we know arities are the same, so skip directly to the terminal case */
-									instanceClass = instanceClassLeafComponent;
-									castClass = castClassLeafComponent;
-									didRetry = true;
-									goto retry;
+								if (!(!J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(instanceClass)
```

Comment:
I prefer 
```
if (J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(instanceClass) 
    || !J9_IS_J9ARRAYCLASS_NULL_RESTRICTED(castClass)
) { 
```

The code should be inside `#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)`